File: operators/grpc_operators/server/grpc_application.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "grpc_application.hpp"

namespace holoscan::ops {

HoloscanGrpcApplication::HoloscanGrpcApplication(
    std::queue<std::shared_ptr<nvidia::gxf::Entity>>& incoming_request_queue,
    std::queue<std::shared_ptr<EntityResponse>>& outgoing_response_queue) {
  request_queue = make_resource<ConditionVariableQueue<std::shared_ptr<nvidia::gxf::Entity>>>(
      "request_queue", incoming_request_queue);
  response_queue = make_resource<ConditionVariableQueue<std::shared_ptr<EntityResponse>>>(
      "response_queue", outgoing_response_queue);
  streaming_enabled = make_condition<BooleanCondition>("streaming_enabled");
}

void HoloscanGrpcApplication::compose() {
  grpc_request_op = make_operator<GrpcServerRequestOp>(
      "grpc_request_op",
      streaming_enabled,
      make_condition<PeriodicCondition>("periodic-condition",
                                        Arg("recess_period") = std::string("60hz")),
      Arg("request_queue") = request_queue,
      Arg("allocator") = make_resource<UnboundedAllocator>("pool"),
      from_config("grpc_server"));

  grpc_response_op = make_operator<GrpcServerResponseOp>(
      "grpc_response_op", streaming_enabled, Arg("response_queue") = response_queue);
}

void HoloscanGrpcApplication::set_scheduler(const std::string& config_name) {
  scheduler(
      make_scheduler<holoscan::EventBasedScheduler>("event-scheduler", from_config(config_name)));
}

void HoloscanGrpcApplication::set_data_path(const std::string path) {
  data_path = path;
}

void HoloscanGrpcApplication::enqueue_request(const EntityRequest& request) {
  auto gxf_allocator = nvidia::gxf::Handle<nvidia::gxf::Allocator>::Create(
      executor().context(), grpc_request_op->allocator()->gxf_cid());
  auto out_message = nvidia::gxf::Entity::New(executor().context());

  holoscan::ops::TensorProto::entity_request_to_tensor(
      &request, out_message.value(), gxf_allocator.value(), grpc_request_op->get_cuda_stream());

  auto entity = std::make_shared<nvidia::gxf::Entity>(out_message.value());
  request_queue->push(entity);
}

void HoloscanGrpcApplication::enqueue_response(std::shared_ptr<EntityResponse> response) {
  response_queue->push(response);
}

bool HoloscanGrpcApplication::is_response_available() {
  return !response_queue->empty();
}

std::shared_ptr<EntityResponse> HoloscanGrpcApplication::dequeue_response() {
  return response_queue->pop();
}

const uint32_t HoloscanGrpcApplication::rpc_timeout() {
  return grpc_request_op->rpc_timeout();
}

void HoloscanGrpcApplication::start_streaming() {
  streaming_enabled->enable_tick();
}

void HoloscanGrpcApplication::stop_streaming() {
  streaming_enabled->disable_tick();
}

}  // namespace holoscan::ops

File: operators/grpc_operators/python/server/grpc_application.py
# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from queue import Queue
from typing import Optional

from holoscan.core import Application
from holoscan.resources import CudaStreamPool, UnboundedAllocator
from holoscan.schedulers import EventBasedScheduler

from holohub.grpc_operators import holoscan_pb2
from operators.grpc_operators.python.common.conditional_variable_queue import ConditionVariableQueue
from operators.grpc_operators.python.server.grpc_server_request import GrpcServerRequestOp
from operators.grpc_operators.python.server.grpc_server_response import GrpcServerResponseOp


class HoloscanGrpcApplication(Application):
    @property
    def composed(self) -> bool:
        return self._composed

    @composed.setter
    def composed(self, value: bool):
        self._composed = value

    @property
    def data_path(self) -> str:
        return self._data_path

    @data_path.setter
    def data_path(self, path: str):
        self._data_path = path

    @property
    def rpc_timeout(self) -> int:
        return self._grpc_request_op.rpc_timeout

    @property
    def grpc_request_op(self) -> GrpcServerRequestOp:
        return self._grpc_request_op

    @property
    def grpc_response_op(self) -> GrpcServerResponseOp:
        return self._grpc_response_op

    @property
    def request_queue(self) -> ConditionVariableQueue:
        return self._request_queue

    @property
    def response_queue(self) -> ConditionVariableQueue:
        return self._response_queue

    def __init__(self, incoming_request_queue: Queue, outgoing_response_queue: Queue):
        super().__init__()
        self.name: str = "gRPC Application"
        self._grpc_request_op: Optional[GrpcServerRequestOp] = None
        self._grpc_response_op: Optional[GrpcServerResponseOp] = None
        self._composed: bool = False
        self._request_queue: ConditionVariableQueue = ConditionVariableQueue(
            self, name="incoming_request_queue", queue=incoming_request_queue
        )
        self._response_queue = ConditionVariableQueue(
            self, name="outgoing_response_queue", queue=outgoing_response_queue
        )
        self.cuda_stream_pool: CudaStreamPool = CudaStreamPool(
            self,
            name="cuda_stream",
            dev_id=0,
            stream_flags=0,
            stream_priority=0,
            reserved_size=1,
            max_size=5,
        )
        self.data_path = ""

    def compose(self):
        self._grpc_request_op = GrpcServerRequestOp(
            self,
            name="grpc_request_op",
            request_queue=self.request_queue,
            allocator=UnboundedAllocator(self, name="pool"),
            cuda_stream_pool=self.cuda_stream_pool,
            **self.kwargs("grpc_server"),
        )
        self._grpc_response_op = GrpcServerResponseOp(
            self,
            name="grpc_response_op",
            response_queue=self.response_queue,
        )

    def set_scheduler(self, config_name: str):
        self.scheduler(EventBasedScheduler(self, "scheduler", **self.kwargs(config_name)))

    def enqueue_request(self, request: holoscan_pb2.EntityRequest):
        self.request_queue.push(request)

    def enqueue_response(self, response: holoscan_pb2.EntityResponse):
        self.response_queue.push(response)

    def is_response_available(self) -> bool:
        return not self.response_queue.empty()

    def dequeue_response(self) -> holoscan_pb2.EntityResponse:
        return self.response_queue.pop()
